src/framework-ondata.js

```diff
@@ -1,18 +1,298 @@
- * @param socketIndex (optional) this will be set only for ports configured for
- *        multiple inputs (multi/noflo addressable).
+ * @param payload data coming into the port
+ * @param socketIndex socket index if the port is a multiple input (addressable) 
  */
 module.exports = function( payload, socketIndex ) {
 
-     return;
+     var portName = this.name;
+
+     var vnid = payload.vnid || '';
+     var vni = this.nodeInstance.vni(vnid);
+
+     var inputState = (payload.vnid) ? payload : stateFactory( vnid, payload );
+     vni.inputStates( portName, socketIndex, inputState );
+
+     if ( shouldRunUpdater( vni ) ) { 
+
+         // Save vars we will need in the promise where the context is different
+         var outputPorts = this.nodeInstance.outPorts;
+
+         if ( _.isUndefined( this.nodeInstance.wrapper.fRunUpdater ) || 
+             ! _.isFunction( this.nodeInstance.wrapper.fRunUpdater ) ) { 
+
+             // Don't have a wrapper fRunUpdater 
+             throw Error( 'No wrapper fRunUpdater function found!  Cannot run updater.' );
+
+         } else { 
+             // TODO: revisit this logic later, per rdf-pipeline/noflo-rdf-pipeline#35
+
+             // save data used to see if we have a different state after running updater
+             var lastErrorState = _.clone( vni.errorState() );  // Shallow clone 
+             var lastOutputLm = vni.outputState().lm;
+ 
+             clearStateData( vni.errorState ); // clear last error state
+
+             // Save wrapper to make it accessible from the Promise
+             var wrapper = this.nodeInstance.wrapper;
+
+             new Promise(function( resolve ) { 
+ 
+                 // Execute fRunUpdater which will also execute the updater
+                 resolve( wrapper.fRunUpdater.call( vni ) );
+
+             }).then( function() { 
+                 // fRunUpdater/Updater success path
+             
+                 // Returned OK, but updater could have set an error - check for that
+                 if ( stateChange( vni.errorState, lastErrorState ) ) {
+                    setOutputErrorFlag( vni );   
+                    lastOutputLm = undefined;  // got a new error so we will force sending output
+                 }
+
+                 // If not new output or error state data, send it downstream or log it to the console
+                 handleOutput( outputPorts.output, lastOutputLm, vni.outputState() );
+                 handleError( outputPorts.error,  lastErrorState.lm, vni.errorState() );
+
+             }, function( rejected ) { 
+                 // fRunUpdater/updater failed 
+                 if ( isInitState( vni.errorState ) ) { 
+                     // fRunUpdater/updater has not already set an error state - use the rejected info
+                     changeStateData( vni.errorState, rejected );
+                 }
+
+                 setOutputErrorFlag( vni, true ); 
+
+                 stateChange( vni.outputState, lastOutputLm );
```

Comment:
Since stateChange will not use the second parameter (if it's not an object) better to use a static value here, like false or undefined, or better yet don't include a second parameter.